### ISSUE TITLE
CS-Script cannot generate on-fly VS project

### ISSUE THREAD

**Issue by KG2501** [oleg-shilo/cs-script#200]
Hello,

I already read #84 and have done what you described there, but on my windows 10 enterprise with visual Studio 2019 Enterprise I still get the error, when try "New Script" or "Open Script".
Is it possible to get any detailed log out put for the error? Maybe permission restrictions from windows or Visual Studio?
On another machine with a similar configuration it works fine.

Maybe you have an idea.

Thanks,
Kai

![Error](https://user-images.githubusercontent.com/49551808/88079922-a7d08580-cb7e-11ea-8d55-c4b66aa12bf8.png)

**Comment by oleg-shilo** [comment]
The most reliable way to investigate is to run it under debuger. But of course it's not necessarily the most practical :o)  

Indeed logging is something that would help to address it and after we solve your problem I will release the update with error logging. But right now let's see what is wrong in your case.

Below is the the complete implementation of the "New Script" use-case.

Try to execute and debug this routine in a simple test console app. Let me know what is failing and I will update the algorithm to address the problem.

I think it might be something wrong with one of the `%CSSCRIPT_DIR%\Lib\debugVS1?.0.cs` paths.

```C#
static void Test()
{
    string scriptFile = CSScriptViewModel.GenerateNewScriptVSSolution();
    string solutionFile = CSScriptViewModel.GenerateScriptVSSolution(scriptFile);
   // VS is going to open `solutionFile`
}

static string GenerateNewScriptVSSolution()
{
    string cssDir = Environment.GetEnvironmentVariable("CSSCRIPT_DIR");

    if (cssDir == null)
        throw new Exception("CS-Script is not installed");

    string srcFile = Path.Combine(cssDir, "lib\\new_script.template");

    var destDir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), "CS-Scripts");
    var destFile = Path.Combine(destDir, "new script.cs");

    if (!Directory.Exists(destDir))
        Directory.CreateDirectory(destDir);

    int i = 0;
    while (File.Exists(destFile))
    {
        i++;
        destFile = Path.Combine(destDir, string.Format("new script ({0}).cs", i));

        if (i > 20)
            throw new Exception("Please cleanup '" + destDir + "' folder");
    }

    File.Copy(srcFile, destFile);

    return destFile;
}

static string GenerateScriptVSSolution(string scriptFile)
{
    string cssDir = Environment.GetEnvironmentVariable("CSSCRIPT_DIR");

    if (cssDir == null)
        throw new Exception("CS-Script is not installed");

    var debug_launcher = FindDebugLauncher();
    if (debug_launcher == null)
        throw new Exception("CS-Script cannot find appropriate debug launcher. Ensure you installed the latest version of CS-Script.");

    string output = RunApp(Path.Combine(cssDir, "cscs.exe"), "/dbg /nl " + debug_launcher + " /noide \"" + scriptFile + "\"");

    if (output.StartsWith("Solution File:"))
    {
        return output.Replace("Solution File:", "").Trim();
    }
    else
    {
        throw new Exception("CS-Script cannot generate on-fly VS project");
    }
}

static string FindDebugLauncher()
{
    if (File.Exists(Environment.ExpandEnvironmentVariables(@"%CSSCRIPT_DIR%\Lib\debugVS15.0.cs")))
        return "debugVS15.0.cs";
    if (File.Exists(Environment.ExpandEnvironmentVariables(@"%CSSCRIPT_DIR%\Lib\debugVS14.0.cs")))
        return "debugVS14.0.cs";
    if (File.Exists(Environment.ExpandEnvironmentVariables(@"%CSSCRIPT_DIR%\Lib\debugVS13.0.cs")))
        return "debugVS13.0.cs";
    else
        return null;
}

istatic string RunApp(string app, string args)
{
    var output = new StringBuilder();

    using (var p = new System.Diagnostics.Process())
    {
        p.StartInfo.FileName = app;
        p.StartInfo.Arguments = args;
        p.StartInfo.UseShellExecute = false;
        p.StartInfo.RedirectStandardOutput = true;
        p.StartInfo.CreateNoWindow = true;
        p.Start();

        string line = null;
        while (null != (line = p.StandardOutput.ReadLine()))
        {
            output.AppendLine(line);
        }
        p.WaitForExit();
        return output.ToString();
    }
}

**Comment by KG2501** [comment]
Thanks for the test code. I ran it and I found the issue. For whatever reason had someone modified the CSSCRIPT_INC by adding an additional path on the machine where I was testing. So when the script debugVS15.0.cs is executed it loads from the css_config.xml the <searchDirs> which includes %CSSCRIPT_INC%. This searchdirs are afterwards processed separately and when try to get the full path for %CSSCRIPT_INC% with the following command: 
  `string dir = Path.GetFullPath(Environment.ExpandEnvironmentVariables(dirItem)); (debugVS15.0.cs, line 361)`
the GetFullPath throws an exception, because %CSSCRIPT_INC% contains multiple path and not just one.

I removed the additional path so that %CSSCRIPT_INC% only contains the original path "C:\ProgramData\CS-Script\inc" and it works.
Sorry for the confusion.
